libgnucash/engine/Transaction.c

```diff
@@ -1805,6 +1804,10 @@ xaccTransRollbackEdit (Transaction *trans)
         }
     }
     g_list_free(slist);
+
+    // orig->splits may still have duped splits so free them
+    for (node = orig->splits; node; node = node->next)
+        xaccFreeSplit(node->data);
     g_list_free(orig->splits);
```

Comment:
Why not `g_list_free_full(orig->splits, xaccFreeSplit)`?

That said I don't see any condition in `xaccTransRollbackEdit` that skips splits that aren't changed. The problem with this is that the splits that are freed aren't removed from the list so this will cause double-frees. Fortunately `xaccFreeSplit` tries to detect this and escape; unfortunately it does it in a way that's incompatible with GObject's way of marking freed objects so it can still crash.

Splits are leaked only if the edited transaction has fewer splits than the clone.

Reply:
> Why not `g_list_free_full(orig->splits, xaccFreeSplit)`?

Did not think of it...
I am a bit confused about the removal from list, what list?
Mind you it does say at the start of this function it is 'terribly complex'.
I have added some print statements starting '=' to the rollback function and on startup I am seeing this...

dupe_trans called, from 0x558a576b8610, to 0x558a56ca4ba0
xaccDupeSplit called, 0x558a576d0660
xaccDupeSplit called, 0x558a576d08c0
xaccDupeSplit called, 0x558a576d09f0
xaccDupeSplit called, 0x558a576d0530
= xaccTransRollbackEdit called, 0x558a576b8610
= xaccTransRollbackEdit called1, split s is 0x558a576baa10
= xaccTransRollbackEdit called1, split s is 0x558a576bab40
= xaccTransRollbackEdit called1, split s is 0x558a576bac70
= xaccTransRollbackEdit called1, split s is 0x558a576ba8e0
= my free orig->splits list 0x558a576d0660
= my free orig->splits list 0x558a576d08c0
= my free orig->splits list 0x558a576d09f0
= my free orig->splits list 0x558a576d0530
  xaccFreeTransaction called, 0x558a56ca4ba0


As you can see for this duped transaction the 4 orig splits are not freed and in total equates to 49 splits which is what I was trying to free. From what I understand is happening in the for loop, it is looking at the transaction slist and iterating over it and if the split is not dirty it continues. So the corresponding split in the orig->slist is not freed.

Reply:
OIC, I was looking at the flow the wrong way. In fact it doesn't ever free any of the cloned splits. I also see where it skips if the original (i.e. not cloned) split isn't marked dirty.

I think from the memory leak standpoint your change is good enough, but the function's assumptions leaves a lot of room for it to do the wrong thing. I guess we're lucky that those cases are rare in real-life usage. Oh well. In the future we'll just discard the edited transaction and get a new instance from the database.